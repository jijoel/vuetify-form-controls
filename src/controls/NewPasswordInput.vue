<template>
<password-input
  :value="value"
  v-bind="$attrs"
  v-on="listeners"
  @input="input"
  @blur="blur"
  :label="label"
  :hint="message"
  :persistentHint="true"
  :loading="loading"
>
  <v-progress-linear
    slot="progress"
    :value="progress"
    :color="color"
    height="2"
  ></v-progress-linear>
</password-input>
</template>

<script>
// These are external scripts to run to check password strength
// They will be loaded when the control is initialized
//
// zxcvbn - check password as it is being entered
// hibp - has the enterd password been pwned? How often?
const strengthCheckers = {
  zxcvbn: 'https://cdnjs.cloudflare.com/ajax/libs/zxcvbn/4.4.2/zxcvbn.js',
  hibp: 'https://unpkg.com/hibp@7.3.0/dist/hibp.min.js',
}
import ChildControl from '../mixins/ChildControl'
import PasswordInput from './PasswordInput'

export default {

  name: 'NewPasswordInput',

  mixins: [ ChildControl ],
  components: { 'password-input': PasswordInput },

  props: {
    label: {
      type: String,
      default: 'New Password'
    },
  },

  data: () => ({
    password_pwned: null,
    password_strength: {
      score: 0,
      feedback: {warning: null, suggestions: []},
    },
  }),

  created() {
    this.loadPasswordStrengthCheckers()
  },

  computed: {
    progress() {
      return this.password_strength.score * 25
    },
    color () {
      return ['error', 'warning', 'success'][Math.floor(this.progress / 40)]
    },
    loading() {
      return !! (this.value && this.value.length)
    },
    message() {
      var message = ''
      if (this.password_pwned)
        message += this.password_pwned

      if (this.password_strength.feedback.warning)
        message += '⚠️ ' + this.password_strength.feedback.warning + '. '

      if (this.password_strength.feedback.suggestions.length)
        message += '🤔 Suggestions: '
          + this.password_strength.feedback.suggestions.join(' ') + '. '

      return message.replace('..', '.')
    }
  },

  methods: {
    blur() {
      this.checkPasswordPwned()
      this.$emit('blur')
    },
    input(value) {
      this.checkPasswordStrength(value)
      this.$emit('input', value)
    },
    loadPasswordStrengthCheckers() {
      Object.keys(strengthCheckers).filter((key) => {
        return (typeof window[key] !== 'function')
      }).map((key) => {
        var tag = document.createElement('script')
        tag.src = strengthCheckers[key]
        document.body.appendChild(tag)
      })
    },
    checkPasswordStrength(value) {
      this.password_pwned = null
      this.password_strength = (!!value && typeof zxcvbn === 'function')
        ? zxcvbn(value)
        : {score: 0, feedback: {warning: null, suggestions: []}}
    },
    checkPasswordPwned() {
      if (typeof hibp !== 'object') {
        console.log('hibp service not found')
        return
      }
      if (!this.value) return

      hibp.pwnedPassword(this.value)
      .then(n => {
        if (n === 0)
          return this.password_pwned = null

        const pwned = [
          {n:10, b:'⚠️', f:'some', s:'questionable'},
          {n:100, b:'⚠️', f:'dozens of', s:'not secure'},
          {n:1000, b:'⚠️', f:'hundreds of', s:'highly insecure'},
          {n:100000, b:'🛑', f:'thousands of', s:'dangerously insecure'},
          {n:Number.MAX_SAFE_INTEGER, b:'🛑', f:'over 100,000', s:'very dangerous to use'},
        ].find(cur => n <= cur.n)

        this.password_pwned = `${pwned.b} Warning: This password has been exposed in ${pwned.f} data breaches. It is ${pwned.s}. Please consider using a different password. `
      })
    }
  },

}
</script>
