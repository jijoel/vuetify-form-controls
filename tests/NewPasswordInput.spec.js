import './setup'
import { common } from './common'
import { shallowMount } from '@vue/test-utils'
import NewPasswordInput from '../src/controls/NewPasswordInput'

describe('NewPasswordInput', () => {

  const zResponse = {
    score: 0,
    feedback: {
      warning: 'x',
      suggestions: ['foo']
    }
  }

  describe('it runs all common tests', () => {
    global.zxcvbn = jest.fn(() => zResponse)
    global.hibp = jest.fn()

    common(NewPasswordInput, { sync: false })
  })

  test('it loads external scripts (zxcvbn & hibp)', () => {
    global.zxcvbn = null
    global.hibp = null
    document.body.appendChild = jest.fn()

    const wrapper = shallowMount(NewPasswordInput)

    const sources = document.body.appendChild.mock.calls.reduce((acc, el) => {
      return acc.concat(el[0].src)
    }, []).join(' ')

    const scripts = ['zxcvbn', 'hibp']
    scripts.map(script => {
      expect(sources).toContain(script)
    })
  })

  // TODO: use mock object for zxcvbn
  // Skip this test for now; the function does not
  // load the correct mock object
  test.skip('it shows progress bar and color based on password strength', () => {
    global.zxcvbn = jest.fn()

    const wrapper = shallowMount(NewPasswordInput)
    wrapper.setValue('x')

    zxcvbn.mockReturnValue(Object.assign(zResponse, {score:5}))
    wrapper.trigger('input', 'x')
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.color).toBe('success')
    expect(wrapper.vm.progress).toBe(100)

    zxcvbn.mockReturnValue(Object.assign(zResponse, {score:3}))
    wrapper.trigger('input', 'x')
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.color).toBe('warning')
    expect(wrapper.vm.progress).toBe(75)

    zxcvbn.mockReturnValue(Object.assign(zResponse, {score:2}))
    wrapper.trigger('input', 'x')
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.color).toBe('warning')
    expect(wrapper.vm.progress).toBe(50)

    zxcvbn.mockReturnValue(Object.assign(zResponse, {score:0}))
    wrapper.trigger('input', 'x')
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.color).toBe('error')
    expect(wrapper.vm.progress).toBe(0)
  })

})
