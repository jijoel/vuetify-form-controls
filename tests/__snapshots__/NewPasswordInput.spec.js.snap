// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPasswordInput it runs all common tests is rendered as custom vuetify text field 1`] = `
<password-input-stub
  browserautocomplete="current-password"
  hint=""
  label="New Password"
  name="password"
  persistenthint="true"
  prependicon="vpn_key"
  value=""
>
  <v-progress-linear-stub
    active="true"
    buffervalue="100"
    color="error"
    height="2"
    value="0"
  />
</password-input-stub>
`;

exports[`NewPasswordInput it runs all common tests renders properties sent by parent 1`] = `
<password-input-stub
  :solo="true"
  browserautocomplete="current-password"
  hint=""
  label="New Password"
  name="password"
  persistenthint="true"
  prependicon="vpn_key"
  value=""
>
  <v-progress-linear-stub
    active="true"
    buffervalue="100"
    color="error"
    height="2"
    value="0"
  />
</password-input-stub>
`;
